Userland/Libraries/LibJS/AST.cpp

```diff
@@ -2011,6 +2012,75 @@ Value MemberExpression::execute(Interpreter& interpreter, GlobalObject& global_o
     return reference.get_value(global_object);
 }
 
+void OptionalChain::dump(int indent) const
+{
+    print_indent(indent);
+    outln("{}", class_name());
+    m_base->dump(indent + 1);
+    for (auto& reference : m_references) {
+        reference.visit(
+            [&](Call const& call) {
+                print_indent(indent + 1);
+                outln("Call({})", call.mode == Mode::Optional ? "Optional" : "Not Optional");
+                for (auto& argument : call.arguments)
+                    argument.value->dump(indent + 2);
+            },
+            [&](ComputedReference const& ref) {
+                print_indent(indent + 1);
+                outln("ComputedReference({})", ref.mode == Mode::Optional ? "Optional" : "Not Optional");
+                ref.expression->dump(indent + 2);
+            },
+            [&](MemberReference const& ref) {
+                print_indent(indent + 1);
+                outln("MemberReference({})", ref.mode == Mode::Optional ? "Optional" : "Not Optional");
+                ref.identifier->dump(indent + 2);
+            });
+    }
+}
+
+Value OptionalChain::execute(Interpreter& interpreter, GlobalObject& global_object) const
+{
+    InterpreterNodeScope node_scope { interpreter, *this };
+
+    auto base_reference = m_base->to_reference(interpreter, global_object);
+    auto base = base_reference.get_value(global_object);
+    if (interpreter.exception())
+        return {};
+
+    for (auto& reference : m_references) {
+        auto is_optional = reference.visit([](auto& ref) { return ref.mode; }) == Mode::Optional;
+        if (is_optional && (base.is_undefined() || base.is_null()))
+            return js_undefined();
+
+        base = reference.visit(
+            [&](Call const& call) {
+                auto call_expression = create_ast_node<CallExpression>(source_range(),
+                    create_ast_node<SyntheticValueExpression>(source_range(), base),
+                    call.arguments);
+                return call_expression->execute(interpreter, global_object);
+            },
+            [&](ComputedReference const& ref) {
+                auto member_expression = create_ast_node<MemberExpression>(source_range(),
+                    create_ast_node<SyntheticValueExpression>(source_range(), base),
+                    ref.expression,
+                    true);
+                return member_expression->execute(interpreter, global_object);
+            },
+            [&](MemberReference const& ref) {
+                auto member_expression = create_ast_node<MemberExpression>(source_range(),
+                    create_ast_node<SyntheticValueExpression>(source_range(), base),
+                    ref.identifier,
+                    false);
+                return member_expression->execute(interpreter, global_object);
+            });
```

Comment:
Wouldn't it be easier to make standalone methods which `execute` in MemberExpression then also calls?

Reply:
Oh I just saw it also has CallExpression then this is probably fine.

Reply:
Yeah these are a bit weird, they make a new node on the fly and execute it, but that's just about the only way I could think of to make these work and not copy half of CallExpression/MemberExpression into this.